soroban-env-host/src/host/diagnostics_helper.rs

```diff
@@ -0,0 +1,105 @@
+use soroban_env_common::{
+    xdr::{ContractEventType, Hash, ScVal, ScVec},
+    ConversionError, ScValObjRef, Symbol, TryFromVal, TryIntoVal, VecObject,
+};
+
+use crate::events::{InternalContractEvent, InternalEvent};
+use crate::{budget::AsBudget, Host, HostError, RawVal};
+
+/// None of these functions are metered, which is why they're behind the is_debug check
+impl Host {
+    fn hash_to_scval(&self, hash: Hash) -> Result<ScVal, HostError> {
+        Ok(ScVal::Bytes(self.scbytes_from_hash(&hash)?))
+    }
+
+    fn record_system_debug_contract_event(
+        &self,
+        type_: ContractEventType,
+        topics: VecObject,
+        data: RawVal,
+    ) -> Result<(), HostError> {
+        let ce = InternalContractEvent {
+            type_,
+            contract_id: None,
```

Comment:
This is a system event, so I don't think it makes sense to include the contract_id here. The contract isn't the one emitting the event. I do include the contract_id being invoked in the topics though, so that information is still there. Curious to hear if you agree though. We could populate contract_id instead.

Reply:
I think if A calls B then a system event with contract_id:A, topic "fn_call", and data B is about right. Otherwise you're not emitting _who called_ B, right? A reader has to infer that from the previous contract-activation in the event log.

Reply:
I initially wanted to avoid including information about the calling contract because it gets complex internally due to the current frame and if a parent frame exists. I see that value in have the calling context though, so I added a commit to include this.